controller/k8s/watcher.go

```diff
@@ -36,6 +41,16 @@ func (w *watcher) run() error {
 	initialized := make(chan struct{}, 1)
 	defer close(initialized)
 
+	go func() {
+		for {
+			err := w.watchInit(w.stopCh)
+			if err != nil {
+				log.Errorf("Error establishing watch in [%s watcher]. Retrying", w.resourceType)
+			}
+			time.Sleep(1 * time.Second)
```

Comment:
This doesn't seem great to for a few reasons:

1 I would expect the proxy to take less than 1s to start up, generally, so ideally the initial timeout should be less than 1s.
2. Backoffs intervals should be increasing over time in general. If that is inappropriate here then there should be a comment explaining why.
3. Retrying *forever* seems wrong to me. If we've been doing this for an hour, for example, then can't we be sure that it's failed? Despite failing to work correctly, we'll be seen as running successfully in terms of Kubernetes' pod management utilities (as reported in kubectl, for example), right?

See https://github.com/kubernetes/client-go/blob/2f45bcbf57f0cfde510a33e7d94a4c89955ff17b/util/retry/util.go#L26-L43.

In particular, should we try to use https://github.com/kubernetes/apimachinery/blob/master/pkg/util/wait/wait.go to address these issues?

Reply:
These are good points. The retry for `util.go` looks like is specifically designed for resource conflicts in k8s so it may be difficult to use that for this retry logic.

I like the idea of using the `wait.go` module, I feel that it is more descriptive I would have to try it out to make sure there is no additional level of complexity being added though. On the issue of retrying indefinitely, it is my understanding that the code retries indefinitely within the 30 seconds and then returns an error which eventually will initiate a pod restart. I think kubectl will report the pod  as being in an `Error` state.